core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java

```diff
@@ -540,18 +538,38 @@ public void delete(K key, InternalCacheEntry<K, V> value, RemovalCause cause) {
                   // If we are here it means this is after the eviction has completed - and thus if the
                   // entry is in the container it means the write put it there, so we skip eviction
                   if (dataContainer.containsKey(key)) {
+                     // AT this point we don't know for a fact if this WRITE has caused the entry to still be in the
+                     // DataContainer or if we are still in the eviction synchronized block just before it is removed
```

Comment:
I never meant for the wording to intend it had to be in a synchronized code block, I was trying to be generic, but I can change the wording.

> I think I understood what's going on in this method, but the comments kind of focus on what's going on in this method instead of explaining why we need to do all this stuff. IIUC Caffeine evicted the entry without the orderer, so another might have acquired the orderer before the actual removal and written the entry after the removal (but PassivationCacheLoaderInterceptor.handlePassivationLoad also acquires the orderer, so reads would be blocked). So we need the 2nd containsKey to see if there was a write after the eviction/removal, in which case we don't need to passivate. We don't strictly need the first containsKey, but it allows us to avoid creating a CompletableFuture instance. Did I get it right @wburns? :)

The problem is you have two different "locks" (the actual container write lock and the orderer). Normal operations acquire the orderer then the container write lock. Eviction does the inverse order.

The main issue for this code is if you have a write to a key at the same time it is evicted. In this case the eviction has the container lock, but the write most likely has the orderer. In that case the eviction will take place removing the object and it is no where (while then writing it to the store). But as you mention reads are blocked, so you can't see that state in the cache. But then the write completes (removes nothing from the store) and the eviction finally writes to the store. Then you can end up with a value in both places.

We have two eviction implementations: Caffeine and our off heap. During eviction the former removes the entry after the listener is invoked, where as off heap removes the entry before notifying.

The problem this code fixes is that we don't for a fact know if this callback is invoked in the eviction thread or a different thread, all we know is that the write has completed. The former we hold the key lock for this container, the latter we do not (we still hold the orderer though). The delay is there to guarantee we are checking the container outside of the container lock during eviction.

Once we are outside of the container lock we can see if the entry is in the container to skip passivation, otherwise we have to passivate it, to keep it in the cache.

The extra containsKey is to avoid the lambda as you mentioned if we happen to not be in the lock any longer.

I believe Caffeine evicts in this way to prevent an eviction if the user desired, by throwing an exception.

I am updating the comments to hopefully make this more clear.